### PR TITLE
Add RefactorReq and RefactorDiffEffect (#996)

### PR DESCRIPTION
**Pull request by lxol**
Extend the ensime protocol to request a refactor change as a diff file in a unified output format.

### CODE REVIEW

**Comment by fommil** [comment]
Oh wow, awesome! Is this the continuation of your work from the hack day? I'll try to find some time this weekend to review this properly.

**Comment by lxol** [comment]
Yes, the hack day was a great jump-start into ensime. Thank you!

File: api/src/main/scala/org/ensime/api/incoming.scala [hunk 1]
```diff
@@ -236,6 +236,16 @@ case class PrepareRefactorReq(
 ) extends RpcAnalyserRequest
 
 /**
+ * Responds with a `RefactorFailure` or a `RefactorDiffEffect`.
+ */
+case class DiffRefactorReq(
+  procId: Int,
+  tpe: Symbol, // tpe is ignored but part of the legacy wire format
```
Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: if you're adding new API messages you don't need to satisfy the legacy format

File: api/src/main/scala/org/ensime/api/incoming.scala [hunk 2]
```diff
@@ -236,6 +236,16 @@ case class PrepareRefactorReq(
...
+  params: RefactorDesc,
```
Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: I'd like to see us come up with a cleaner way of doing the refactor params. This legacy object is quite horrible.

File: project/EnsimeBuild.scala [hunk 3]
```diff
@@ -312,7 +312,8 @@ object EnsimeBuild extends Build with JdkResolver {
       "com.typesafe.akka" %% "akka-slf4j" % akkaVersion,
       "org.scala-refactoring" %% "org.scala-refactoring.library" % "0.8.0",
       "commons-lang" % "commons-lang" % "2.6",
-      "commons-io" % "commons-io" % "2.4" % "test,it"
+      "commons-io" % "commons-io" % "2.4" % "test,it",
+      "com.googlecode.java-diff-utils" % "diffutils" % "1.3.0"
```
Comment (fommil) [inline project/EnsimeBuild.scala]: great find!

File: swank/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala [hunk 4]
```diff
@@ -442,6 +442,7 @@ object SwankProtocolResponse {
   }
   // must be after FileEditFormat
   implicit val RefactorEffectFormat = SexpFormat[RefactorEffect]
+  implicit val RefactorDiffEffectFormat = SexpFormat[RefactorDiffEffect]
```
Comment (fommil) [inline swank/src/main/scala/org/ensime/server/protocol/swank/SwankFormats.scala]: I'm impressed that you worked out the changes needed in this file :smile: it's really quite horrible. I'd love to just delete the whole thing, but it requires a big migration on the emacs side.

File: api/src/main/scala/org/ensime/api/incoming.scala (hunk 2 above)
Comment (fommil) [inline api/src/main/scala/org/ensime/api/incoming.scala]: ok, I see now that this gives us a really easy migration path onto diffs, which is the most important thing.

File: core/src/main/scala/org/ensime/util/FileUtil.scala [hunk 5]
```diff
@@ -82,4 +82,28 @@ object FileUtils {
...
+  def writeDiffChanges(changes: List[FileEdit], cs: Charset): Either[Exception, File] = {
+    //TODO: add support for NewFile and DeleteFile
+    val editsByFile = changes.collect { case ed: TextEdit => ed }.groupBy(_.file)
+    try {
+      val diffContents =
+        editsByFile.map {
+          case (file, fileChanges) =>
+            readFile(file, cs) match {
+              case Right(contents) =>
+                FileEditHelper.diffFromTextEdits(fileChanges, contents, "/dev/null", file.getAbsolutePath())
```
Comment (fommil) [inline core/src/main/scala/org/ensime/util/FileUtil.scala]: I'm surprised that this passes on the Windows CI

**Comment by fommil** [comment]
ok, this is really great stuff, thank you so much! Can you please:

1. remove the redundant fields in the incoming/outgoing
2. change the API so that we don't need to pass this `"/dev/null"` string (input/output name could be the same)
3. squash into one commit

and this looks like it's then ready for merge!

**Comment by aemoncannon** [comment]
Awesome work. What do you think about PrepareRefactorReqV2 over DiffRefactorReq? Ditto for RefactorResultV2. Would make the evolution of the protocol easier to see, and maybe could adopt this approach in general. It's good enough for Sqlite :-)

**Comment by aemoncannon** [comment]
Also please bump the protocol version and add an entry in the protocol change log.

**Comment by fommil** [comment]
ah, yes, forgot about that. it's in the `swank` folder if you're interested.

**Comment by lxol** [comment]
@aemoncannon PrepareRefactorReqV2 is supposed to replace PrepareRefactorReq ExecRefactorRec and CancelRefactorReq. Don't you thinks that we should remove "Prepare" prefix?

**Comment by lxol** [comment]
I will make the changes tomorrow. 
Thank you for the review

**Comment by aemoncannon** [comment]
@lxol I see. So the clients are expected to apply the diff -- is that correct? In that case, maybe just 'RefactorReq' , and 'RefactorDiffResult'. What do you think?

**Comment by aemoncannon** [comment]
Er, RefactorDiffEffect

**Comment by lxol** [comment]
@aemoncannon Yes, the editor will be responsible for the edits and undos. 

I will rename DiffRefactorReq to RefactorReq as you suggested

**Comment by fommil** [comment]
can you please force push onto this PR with your changes?

**Comment by lxol** [comment]
done

**Comment by fommil** [comment]
This is a really top notch PR, thank you! Fancy having a look at moving over Emacs?

**Comment by lxol** [comment]
@fommil I already used Emacs to send/receive diff request/responses using a new api. I will try to make  something production ready. Not sure about ETA though.

**Comment by fommil** [comment]
:grinning:

**Comment by aemoncannon** [comment]
Huzzah!

**Comment by fommil** [comment]
@ensime/vim-panel @ensime/atom-panel @ensime/sublime-panel this PR introduces a much better way to perform refactorings. We'd like to move all clients over to this new (simpler) API, can I bother you all to look at doing this? I'd like to remove the legacy APIs in the 1.0 release (yeah yeah, I know its 10 months overdue... but upgrading to Java 7 is becoming more of a priority so it might happen).

It's important as part of our "ENSIME server shouldn't mutate anything" strategy.

**Comment by davegurnell** [comment]
Sorry for the delayed reply... better late than never!

In general this looks great to me. IIUC these are Git-compatible diffs?

Dave

**Comment by lxol** [comment]
@davegurnell the diff file is in a unified diff format and git-apply can use it. I've just tested git-apply and noticed several issued. 
This command complains that the diff format is wrong.
```` sh
git apply -p0 --unsafe-path /private/var/folders/hc/lg9jzhmn5zj4kp4ldl2hfmk00000gn/T/ensime-diff-1519913091103607566.tmp
````
However if I add an empty line to the end of the diff file, git-apply  works perfectly. I will submit a PR to fix that.
Another issue is that using an external command to apply the patch is no different from applying the changes from the ensime-server. We loose the benefit of keeping the edit history.  We are lucky with `diff-mode` in Emacs.

**Comment by davegurnell** [comment]
Ok awesome. I realised reading your code that I don't know enough about
diffs. I wasn't sure what libraries to use to interpret them. What you've
said there sounds great---a universal or git-compatible diff format is
bound to have support in all the various client languages.

**Pull request merged**